### PR TITLE
Search: Improvements for starred dashboard search

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Several improvements related to starred dashboards:
* when searching for starred dashboards, pass dashboard IDs into the search to optimise search performance (it's way faster if dashboard IDs are specified);
* when using starred dashboards, fetch them all in one query rather than issuing one query per dashboard.

**Why do we need this feature?**

We've had reports of starred dashboard queries being slow.

**Special notes for your reviewer**:

I've tested with 20 starred dashboards and it performed fine. However, it might not work well if the user has a really large number of starred dashboards. I'm working with the assumption that a typical user will not have more than 20 starred dashboards.

### CODE REVIEW

**Review by Jguer — APPROVED** [review]
Nice inversion of the search domain

File: pkg/services/guardian/accesscontrol_guardian.go [hunk 1]
```diff
@@ -224,11 +224,19 @@ func (a *AccessControlDashboardGuardian) CanCreate(folderID int64, isFolder bool
 func (a *AccessControlDashboardGuardian) evaluate(evaluator accesscontrol.Evaluator) (bool, error) {
 	ok, err := a.ac.Evaluate(a.ctx, a.user, evaluator)
 	if err != nil {
-		a.log.Debug("Failed to evaluate access control to folder or dashboard", "error", err, "userId", a.user.UserID, "id", a.dashboard.ID)
+		id := 0
+		if a.dashboard != nil {
+			id = int(a.dashboard.ID)
+		}
+		a.log.Debug("Failed to evaluate access control to folder or dashboard", "error", err, "userId", a.user.UserID, "id", id)
```
Comment (IevaVasiljeva) [inline pkg/services/guardian/accesscontrol_guardian.go:231]: This and the following change are not related to the PR. I just noticed that this can panic while testing the PR, so added a small fix. Panic comes from cases when a dashboard is created under the general folder - in that case `a.dashboard` will be `nil`.

**Comment by ryantxu** [comment]
This is fine, but we should really just move to a place where the star service is not intertwined with the other services.  Given the scope, we can simply return all starred UIDs to the frontend, and mark the existence of a star from there.

File: pkg/services/navtree/navtreeimpl/navtree.go [hunk 2]
```diff
@@ -345,22 +345,17 @@ func (s *ServiceImpl) buildStarredItemsNavLinks(c *contextmodel.ReqContext) ([]*
...
+	for id := range starredDashboardResult.UserStars {
+		ids = append(ids, id)
+	}
+	starredDashboards, err := s.dashboardService.GetDashboards(c.Req.Context(), &dashboards.GetDashboardsQuery{DashboardIDs: ids, OrgID: c.OrgID})
+	if err != nil {
+		return nil, err
+	}
+	// Set a loose limit to the first 50 starred dashboards found
+	if len(starredDashboards) > 50 {
+		starredDashboards = starredDashboards[:50]
```
Comment (kalleep) [inline pkg/services/navtree/navtreeimpl/navtree.go:358]: We need to check if we have any ids and only query for dashboards in that case. Now if I have no starred dashboards I get an error and cannot render the page

File: pkg/services/search/service.go [hunk 3]
```diff
@@ -58,10 +58,24 @@ type SearchService struct {
...
+		UserID: query.SignedInUser.UserID,
+	}
+	staredDashIDs, err := s.starService.GetByUser(ctx, &starredQuery)
+	if err != nil {
+		return err
+	}
+
+	// filter by starred dashboard IDs when starred dashboards are requested and no UID or ID filters are specified to improve query performance
+	if query.IsStarred && len(query.DashboardIds) == 0 && len(query.DashboardUIDs) == 0 {
+		for id := range staredDashIDs.UserStars {
```
Comment (kalleep) [inline pkg/services/search/service.go:71]: Nit: We can short circuit the request here if we don't have any starred dashboards

**Review by kalleep — CHANGES_REQUESTED** [review]
We cannot merged until https://github.com/grafana/grafana/pull/64758#discussion_r1136676612 is fixed

File: pkg/services/navtree/navtreeimpl/navtree.go (hunk 2 above)
Comment (Jguer) [reply pkg/services/navtree/navtreeimpl/navtree.go:358]: should we add the ids protection (return empty slice and no error when len(ids) == 0) inside of `dashboardService.GetDashboards` to avoid it happening elsewhere

File: pkg/services/star/api/api.go [hunk 4]
```diff
@@ -52,22 +52,23 @@ func (api *API) GetStars(c *contextmodel.ReqContext) response.Response {
...
-		query := &dashboards.GetDashboardQuery{
-			ID:    dashboardId,
-			OrgID: c.OrgID,
-		}
-		queryResult, err := api.dashboardService.GetDashboard(c.Req.Context(), query)
+	var ids []int64
+	for id := range iuserstars.UserStars {
+		ids = append(ids, id)
+	}
+	starredDashboards, err := api.dashboardService.GetDashboards(c.Req.Context(), &dashboards.GetDashboardsQuery{DashboardIDs: ids, OrgID: c.OrgID})
```
Comment (kalleep) [inline pkg/services/star/api/api.go:62]: Same here, we need to check that we have ids

File: pkg/services/navtree/navtreeimpl/navtree.go (hunk 2 above)
Comment (Jguer) [reply pkg/services/navtree/navtreeimpl/navtree.go]: @IevaVasiljeva sure thing

File: pkg/services/search/service.go (hunk 3 above)
Comment (IevaVasiljeva) [reply pkg/services/search/service.go:71]: Good shout, added the change.

File: pkg/services/navtree/navtreeimpl/navtree.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/navtree/navtreeimpl/navtree.go]: Starred dashboard table doesn't store org ID for the org that the dashboard is in, so it could be the case that the first 50 starred dashboard IDs are from a different org. Therefore I think we need to query for all of them unfortunately.

[the remark this answers is not in the document]
Comment (kalleep) [reply pkg/services/navtree/navtreeimpl/navtree.go]: I see

**Review by kalleep — APPROVED** [review]

File: pkg/services/navtree/navtreeimpl/navtree.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (suntala) [reply pkg/services/navtree/navtreeimpl/navtree.go]: It's been a long day so I might not be looking at this right. These `starredDashboards` are getting fetched with this [GetDashboards](https://github.com/grafana/grafana/blob/f013d37ce965bd5789856113c2e36310da1c45e7/pkg/services/navtree/navtreeimpl/navtree.go#L353) call which wraps around [FindDashboards](https://github.com/grafana/grafana/blob/main/pkg/services/dashboards/database/database.go#L915). That's not the query to starred dashboard table that you're talking about is it?

**Comment by IevaVasiljeva** [comment]
> This is fine, but we should really just move to a place where the star service is not intertwined with the other services. Given the scope, we can simply return all starred UIDs to the frontend, and mark the existence of a star from there.

Thanks for the idea! I think it might require quite a big refactor, but I'll share this with the owners of the star service.
